Update Release.md with latest details (#78285)
pytorch/pytorch: RELEASE.md

  - [Cutting release branches](#cutting-release-branches)
    - [`pytorch/pytorch`](#pytorchpytorch)
    - [`pytorch/builder` / PyTorch domain libraries](#pytorchbuilder--pytorch-domain-libraries)

>     - [Making release branch specific changes for pytorch](#making-release-branch-specific-changes-for-pytorch)

Reviewer: Please properly capitalize PyTorch
```suggestion
    - [Making release branch specific changes for PyTorch](#making-release-branch-specific-changes-for-pytorch)
```

    - [Making release branch specific changes for domain libraries](#making-release-branch-specific-changes-for-domain-libraries)

>   - [Drafting RCs (Release Candidates) for PyTorch and domain libraries](#drafting-rcs-release-candidates-for-pytorchpytorch-and-domain-libraries)

Author: ```suggestion
  - [Drafting RCs (Release Candidates) for PyTorch and domain libraries](#drafting-rcs-release-candidates-for-pytorch-and-domain-libraries)
```

>   - [Drafting RCs (Release Candidates) for `pytorch/pytorch` and domain libraries](#drafting-rcs-release-candidates-for-pytorchpytorch-and-domain-libraries)

Reviewer: Please properly capitalize PyTorch (and perhaps replace domain with ecosystem?
```suggestion
  - [Drafting RCs (Release Candidates) for PyTorch and domain libraries](#drafting-rcs-release-candidates-for-pytorchpytorch-and-domain-libraries)
```

Release branches are typically cut from the branch [`viable/strict`](https://github.com/pytorch/pytorch/tree/viable/strict) as to ensure that tests are passing on the release branch.

There's a convenience script to create release branches from current `viable/strict`. Perform following actions :

> * Perform a fresh clone of pytorch repo using
> ```bash
> git clone git@github.com:pytorch/pytorch.git
> ```

Reviewer: This sounds like an optional step, isn't it? Also, why are you recommending using ssh rather than https access? Later is faster, isn't it?
```suggestion
* Perform a fresh clone of pytorch repo using
```bash
git clone https://github.com:pytorch/pytorch.git
```
```

Author: 1. Basically this step is required if you want to execute next step which is convenience script, which should be executed from the pytorch clone. Executing the script from fork does not work. Hence fresh clone, which also makes sure no other accidental changes are picked up during release.

2.  I don't mind putting https here if it works well for next step, so far I have been always using ssh with keys setup for proper access. Main point here is that it must be pytorch/pytorch repo and not a fork.

Reviewer: Sorry, I don't think I follow. What do you mean by:
> which should be executed from the pytorch clone
If script does not work from forks, then perhaps we need to add note about it here and also add a check for it in the script.

Author: Yes, makes sense we should probably change the script to work from forks properly.  The way I usually work is as follows:

```
(.venv) [atalman@devvm214.ftw0 ~/local/pytorch] git remote -v
origin  git@github.com:atalman/pytorch.git (fetch)
origin  git@github.com:atalman/pytorch.git (push)
upstream        git@github.com:pytorch/pytorch.git (fetch)
upstream        git@github.com:pytorch/pytorch.git (push)
```

The script by default assumes:  ```GIT_REMOTE=${GIT_REMOTE:-origin}``` hence if I execute it from fork it will push it to atalman/pytorch.

Perhaps we should just avoid using the script since it creates unnecessary confusion, all it does:
```
set -x
git checkout "${GIT_REMOTE}/${GIT_BRANCH_TO_CUT_FROM}"
git checkout -b "${branch}"
git push "${GIT_REMOTE}" "${branch}"
```

Maybe just executing these command against correct origin or upstream will be cleaner ?

> * Change to pytorch repo root and run following commands:

Reviewer: Please capitalize PyTorch, also, I'm not sure what `Change to pytorch repo root` means, perhaps it better be rephrased as follows
```suggestion
* Execute following command from PyTorch repository root folder:
```

```bash
DRY_RUN=disabled scripts/release/cut-release-branch.sh
```
This script should create 2 branches:
* `release/{MAJOR}.{MINOR}`
* `orig/release/{MAJOR}.{MINOR}`

### `pytorch/builder` / PyTorch domain libraries

> Important note: Cutting the branch for individual domain libraries is performed a week later after branch cut for the Pytorch and Builder. This is absolutely required to allow sufficient testing time for each of the domain library. Domain libraries branch cut is performed by Domain Library POC.

Reviewer: Not sure if mentioning time limits is useful here. Also, s/cutting/creating/
```suggestion
*Note*:  Release branches for individual domain libraries should be created after first release candidate build of PyTorch is available in staging channels (which happens about a week after PyTorch release branch has been created). This is absolutely required to allow sufficient testing time for each of the domain library. Domain libraries branch cut is performed by Domain Library POC.
```

> Important note: Cutting the branch for individual domain libraries is performed a week later after branch cut for the Pytorch and Builder. This is absolutely required to allow sufficient testing time for each of the domain library. Domain libraries branch cut is performed by Domain Library POC. 

Author: ```suggestion
Important note: Cutting the branch for individual domain libraries is performed a week later after branch cut for the Pytorch and Builder. This is absolutely required to allow sufficient testing time for each of the domain library. Domain libraries branch cut is performed by Domain Library POC.```

Builder branch cut should be performed at the same time as Pytorch core branch cut. Convenience script can also be used domains as well as `pytorch/builder`

> NOTE: RELEASE_VERSION only needs to be specified if version.txt is not available in root directory

```bash
DRY_RUN=disabled GIT_BRANCH_TO_CUT_FROM=main RELEASE_VERSION=1.11 scripts/release/cut-release-branch.sh
```

> ### Making release branch specific changes for pytorch

Reviewer: Please capitalize PyTorch

Author: Done

> ### Making release branch specific changes for pytorch

Author: ```suggestion
### Making release branch specific changes for PyTorch
```

* A release branches should also be created in [`pytorch/xla`](https://github.com/pytorch/xla) and [`pytorch/builder`](https://github.com/pytorch/builder) repos and pinned in `pytorch/pytorch`

These are examples of changes that should be made to the *default* branch after a release branch is cut

* Nightly versions should be updated in all version files to the next MINOR release (i.e. 0.9.0 -> 0.10.0) in the default branch:
  * Example: https://github.com/pytorch/pytorch/pull/77984

### Making release branch specific changes for domain libraries

Domain library branch cut is done a week after branch cut for the `pytorch/pytorch`. The branch cut is performed by the Domain Library POC.
After the branch cut is performed, the Pytorch Dev Infra memeber should be informed of the branch cut and Domain Library specific change is required before Drafting RC for this domain library.

Follow these examples of PR that updates the version and sets RC Candidate upload channel:
* torchvision : https://github.com/pytorch/vision/pull/5400
* torchaudio: https://github.com/pytorch/audio/pull/2210

> ## Drafting RCs (Release Candidates) for `pytorch/pytorch` and domain libraries

Author: ```suggestion
## Drafting RCs (Release Candidates) for PyTorch and domain libraries
```

```bash
DRY_RUN=disabled scripts/release/cut-release-branch.sh
```
This script should create 2 branches:
* `release/{MAJOR}.{MINOR}`
* `orig/release/{MAJOR}.{MINOR}`

### `pytorch/builder` / PyTorch domain libraries

Important note: Cutting the branch for individual domain libraries is performed a week later after branch cut for the Pytorch and Builder. This is absolutely required to allow sufficient testing time for each of the domain library. Domain libraries branch cut is performed by Domain Library POC. 
Builder branch cut should be performed at the same time as Pytorch core branch cut. Convenience script can also be used domains as well as `pytorch/builder`

> NOTE: RELEASE_VERSION only needs to be specified if version.txt is not available in root directory

```bash
DRY_RUN=disabled GIT_BRANCH_TO_CUT_FROM=main RELEASE_VERSION=1.11 scripts/release/cut-release-branch.sh
```

### Making release branch specific changes for pytorch

These are examples of changes that should be made to release branches so that CI / tooling can function normally on
them:

* Update backwards compatibility tests to use RC binaries instead of nightlies

>   * Example: https://github.com/pytorch/pytorch/pull/77983

Author: ```suggestion
  * Example: https://github.com/pytorch/pytorch/pull/77983 and https://github.com/pytorch/pytorch/pull/77986
```

* A release branches should also be created in [`pytorch/xla`](https://github.com/pytorch/xla) and [`pytorch/builder`](https://github.com/pytorch/builder) repos and pinned in `pytorch/pytorch`

>   * Example PRs (GHA): https://github.com/pytorch/pytorch/pull/77983 and https://github.com/pytorch/pytorch/pull/77986

Author: ```suggestion
```

Release branches are typically cut from the branch [`viable/strict`](https://github.com/pytorch/pytorch/tree/viable/strict) as to ensure that tests are passing on the release branch.

There's a convenience script to create release branches from current `viable/strict`. Perform following actions :
* Perform a fresh clone of pytorch repo using
```bash
git clone git@github.com:pytorch/pytorch.git
```

* Change to pytorch repo root and run following commands:
```bash
DRY_RUN=disabled scripts/release/cut-release-branch.sh
```
This script should create 2 branches:
* `release/{MAJOR}.{MINOR}`
* `orig/release/{MAJOR}.{MINOR}`

### `pytorch/builder` / PyTorch domain libraries

Important note: Cutting the branch for individual domain libraries is performed a week later after branch cut for the Pytorch and Builder. This is absolutely required to allow sufficient testing time for each of the domain library. Domain libraries branch cut is performed by Domain Library POC. 

> 

Author: ```suggestion
```

Builder branch cut should be performed at the same time as Pytorch core branch cut. Convenience script can also be used domains as well as `pytorch/builder`

> NOTE: RELEASE_VERSION only needs to be specified if version.txt is not available in root directory

```bash
DRY_RUN=disabled GIT_BRANCH_TO_CUT_FROM=main RELEASE_VERSION=1.11 scripts/release/cut-release-branch.sh
```

### Making release branch specific changes for pytorch

These are examples of changes that should be made to release branches so that CI / tooling can function normally on
them:

* Update backwards compatibility tests to use RC binaries instead of nightlies
  * Example: https://github.com/pytorch/pytorch/pull/77983
* A release branches should also be created in [`pytorch/xla`](https://github.com/pytorch/xla) and [`pytorch/builder`](https://github.com/pytorch/builder) repos and pinned in `pytorch/pytorch`

>   * Example PRs (GHA): https://github.com/pytorch/pytorch/pull/77986 and https://github.com/pytorch/pytorch/pull/77986

Reviewer 2: these are the same PRs

Author: Thanks refactored!

>   * Example PRs (GHA): https://github.com/pytorch/pytorch/pull/77986 and https://github.com/pytorch/pytorch/pull/77986

Author: ```suggestion
  * Example PRs (GHA): https://github.com/pytorch/pytorch/pull/77983 and https://github.com/pytorch/pytorch/pull/77986
```

An example of this would look like:
```
v1.12.0-rc1
```
You can use following commands to perform tag from pytorch core repo (not fork):
* Checkout and validate the repo history before tagging
```
git checkout release/1.12
git log --oneline
```
* Perform tag and push it to github (this will trigger the binary release build)
```
git tag -f  v1.12.0-rc2
git push origin  v1.12.0-rc2
```

Pushing a release candidate should trigger the `binary_builds` workflow within CircleCI using [`pytorch/pytorch-probot`](https://github.com/pytorch/pytorch-probot)'s [`trigger-circleci-workflows`](trigger-circleci-workflows) functionality.

This trigger functionality is configured here: [`pytorch-circleci-labels.yml`](https://github.com/pytorch/pytorch/blob/master/.github/pytorch-circleci-labels.yml)

To view the state of the release build, please navigate to [HUD](https://hud.pytorch.org/hud/pytorch/pytorch/release%2F1.12). and make sure all binary builds are successful.

> 

Author: ```suggestion
```

An example of this would look like:
```
v1.12.0-rc1
```

> 

Author: ```suggestion
```
